astra/src/test/java/com/slack/astra/server/ManagerApiGrpcTest.java

```diff
@@ -452,29 +460,805 @@ public void shouldUpdatePartitionAssignments() {
+            .setThroughputBytes(8000000)
+            .setRequireDedicatedPartition(false)
+            .build());
+
+    await().until(() -> datasetMetadataStore.getSync(datasetName1).getThroughputBytes() == 8000000);
+
+    assertThat(getDatasetMetadataGRPC(datasetName1))
+        .has(throughputBytesOf(8000000))
+        .has(partitionConfigsSizeOf(3))
+        .has(partitionsForIndexOf(0, List.of("1", "2", "3")))
+        .has(partitionsForIndexOf(1, List.of("3", "4", "5")))
+        .has(partitionsForIndexOf(2, List.of("3", "4")))
+        .has(oldPartitionConfigWithIndexOf(0))
+        .has(oldPartitionConfigWithIndexOf(1))
+        .has(latestPartitionConfigWithIndexOf(2, nowMs));
+
+    // unchanged
+    assertThat(getDatasetMetadataGRPC(datasetName2))
+        .has(throughputBytesOf(2000000))
+        .has(partitionConfigsSizeOf(1))
+        .has(partitionsForIndexOf(0, List.of("1", "2")))
+        .has(latestPartitionConfigWithIndexOf(0, nowMs));
+
+    assertThat(getPartitionMetadata("1", "2")).have(sharedPartitionsWithCapacity(1000000));
+    assertThat(getPartitionMetadata("3", "4")).have(sharedPartitionsWithCapacity(4000000));
+    assertThat(getPartitionMetadata("5")).have(sharedPartitionsWithCapacity(0));
+
+    // fourth Assignment: move dataset 1 to shared partitions, increase throughput to 12000000
+    nowMs = Instant.now().toEpochMilli();
+    managerApiStub.updatePartitionAssignment(
+        ManagerApi.UpdatePartitionAssignmentRequest.newBuilder()
+            .setName(datasetName1)
+            .setThroughputBytes(12000000)
+            .setRequireDedicatedPartition(false)
+            .build());
+
+    await()
+        .until(() -> datasetMetadataStore.getSync(datasetName1).getThroughputBytes() == 12000000);
+
+    assertThat(getDatasetMetadataGRPC(datasetName1))
+        .has(throughputBytesOf(12000000))
+        .has(partitionConfigsSizeOf(4))
+        .has(partitionsForIndexOf(0, List.of("1", "2", "3")))
+        .has(partitionsForIndexOf(1, List.of("3", "4", "5")))
+        .has(partitionsForIndexOf(2, List.of("3", "4")))
+        .has(partitionsForIndexOf(3, List.of("1", "3", "4")))
+        .has(oldPartitionConfigWithIndexOf(0))
+        .has(oldPartitionConfigWithIndexOf(1))
+        .has(oldPartitionConfigWithIndexOf(2))
+        .has(latestPartitionConfigWithIndexOf(3, nowMs));
+
+    // unchanged
+    assertThat(getDatasetMetadataGRPC(datasetName2))
+        .has(throughputBytesOf(2000000))
+        .has(partitionConfigsSizeOf(1))
+        .has(partitionsForIndexOf(0, List.of("1", "2")))
+        .has(latestPartitionConfigWithIndexOf(0, nowMs));
+
+    assertThat(getPartitionMetadata("1")).have(sharedPartitionsWithCapacity(5000000));
+    assertThat(getPartitionMetadata("2")).have(sharedPartitionsWithCapacity(1000000));
+    assertThat(getPartitionMetadata("3", "4")).have(sharedPartitionsWithCapacity(4000000));
+    assertThat(getPartitionMetadata("5")).have(sharedPartitionsWithCapacity(0));
+  }
+
+  @Test
```

Comment:
Add a unit test where partitions can have variable max capacity. That may help surface some more bugs.